cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -140,3 +144,62 @@ func WriteStaticPodToDisk(componentName, manifestDir string, pod v1.Pod) error {
 
 	return nil
 }
+
+// GetProbeAddress returns an IP address or 127.0.0.1 to use for liveness probes
+// in static pod manifests. For etcd, GetProbeAddress will use the first
+// listen-client-url if multiple urls are provided. If the listen-client-url
+// contains a hostname instead of an IP address, GetProbeAddress will use the
+// local resolver for name resolution or 127.0.0.1 if the name can not be resolved.
+func GetProbeAddress(cfg *kubeadmapi.MasterConfiguration, component string) string {
+	switch {
+	case component == kubeadmconstants.KubeAPIServer:
+		if cfg.API.AdvertiseAddress != "" {
+			return cfg.API.AdvertiseAddress
+		}
+	case component == kubeadmconstants.KubeScheduler:
+		if cfg.SchedulerExtraArgs != nil {
+			addr, exists := cfg.SchedulerExtraArgs["address"]
+			if exists {
+				return addr
+			}
+		}
+	case component == kubeadmconstants.KubeControllerManager:
+		if cfg.ControllerManagerExtraArgs != nil {
+			addr, exists := cfg.ControllerManagerExtraArgs["address"]
+			if exists {
+				return addr
+			}
+		}
+	case component == kubeadmconstants.Etcd:
+		if cfg.Etcd.ExtraArgs != nil {
+			arg, exists := cfg.Etcd.ExtraArgs["listen-client-urls"]
+			if exists {
+				parsedURL, err := url.Parse(arg)
+				if err != nil || parsedURL.Hostname() == "" {
+					break
+				}
+				if ip := net.ParseIP(parsedURL.Hostname()); ip != nil {
+					return ip.String()
+				}
+				addrs, err := net.LookupIP(parsedURL.Hostname())
+				if err != nil {
+					break
+				}
+				var ip net.IP
+				for _, addr := range addrs {
+					if addr.To4() != nil {
+						ip = addr
```

Comment:
the result here is that ipv4 is always prioritised over ipv6 if a component uses hostnames. we should add a test to make this behaviour explicit.

Reply:
That is correct. Preferring v4 over v6 is the approach that is taken elsewhere in the project. Since we pass a hostname to `net.LookupIP`, it can be challenging to test the behavior in unit tests b/c it depends on the layout of `/etc/hosts`. Let me know if you have suggestions.

Reply:
seems like you could `return addr.To4().String()` here? to make the code shorter

Reply:
ping @danehans